README.md
[![Build Status](https://travis-ci.org/wmateam/curling.svg?branch=master)](https://travis-ci.org/wmateam/curling)
[![Total Downloads](https://poser.pugx.org/wmateam/curling/downloads)](https://packagist.org/packages/wmateam/curling)
[![Latest Stable Version](https://poser.pugx.org/wmateam/curling/v/stable)](//packagist.org/packages/wmateam/curling)
[![Latest Unstable Version](https://poser.pugx.org/wmateam/curling/v/unstable)](//packagist.org/packages/wmateam/curling)
[![License](https://poser.pugx.org/wmateam/curling/license)](https://packagist.org/packages/wmateam/curling)
# Curling
CURL request Handler for PHP

Support :

* GET
* POST
* PUT
* PATCH
* DELETE

You can use proxy on your request.

This library support Authentication.
# [Wiki and Docs](https://github.com/wmateam/curling/wiki)

CHANGELOG.md
# Curing Changelog
## 2016-05-24, Version 0.4.0

You can send request as a json or text

## 2016-01-20, Version 0.7.0

Rename `HttpRequest` to `CurlRequest`

Rename `HttpResponse` to `CurlResponse`

tests/RequestTest.php
<?php

/**
 * User: wmateam
 * Date: 5/15/16
 * Time: 1:45 PM
 */

use wmateam\curling\CurlRequest;

class RequestTest extends PHPUnit_Framework_TestCase
{
    private $url = 'https://httpbin.org/';
    private $params = array('foo' => 'bar', 'baz' => 'boom', 'cow' => 'milk');
    private $jsonParams = '{"array":[{"cow":"milk"}],"baz":"boom","foo":"bar"}';
    private $headers = 'agent: wmateam\curling';

    public function testGetRequest()
    {
        echo __METHOD__ . "\n";
        $r = new CurlRequest($this->url . 'get');
        $r->setQueryString($this->params);
        $r->setHeader($this->headers);
        $data = $r->get();
        $this->assertEquals(200, $data->getStatusCode());
    }

    public function testPostRequest()
    {
        echo __METHOD__ . "\n";
        $r = new CurlRequest($this->url . 'post');
        $r->setQueryString($this->params);
        $r->setHeader($this->headers);
        $data = $r->post($this->params);
        $this->assertEquals(200, $data->getStatusCode());
    }

    public function testPostRawDataRequest()
    {
        echo __METHOD__ . "\n";
        $r = new CurlRequest($this->url . 'post');
        $r->setQueryString($this->params);
        $r->setHeader($this->headers);
        $data = $r->post($this->jsonParams, CurlRequest::RAW_DATA);
        $this->assertEquals(true, json_decode($data->getBody(), true)['json'] != '');
        $this->assertEquals(200, $data->getStatusCode());
    }

    public function testPutRequest()
    {
        echo __METHOD__ . "\n";
        $r = new CurlRequest($this->url . 'put');
        $r->setQueryString($this->params);
        $r->setHeader($this->headers);
        $data = $r->put($this->params);
        $this->assertEquals(200, $data->getStatusCode());
    }

    public function testPatchRequest()
    {
        echo __METHOD__ . "\n";
        $r = new CurlRequest($this->url . 'patch');
        $r->setQueryString($this->params);
        $r->setHeader($this->headers);
        $data = $r->patch($this->params);
        $this->assertEquals(200, $data->getStatusCode());
    }

    public function testDeleteRequest()
    {
        echo __METHOD__ . "\n";
        $r = new CurlRequest($this->url . 'delete');
        $r->setQueryString($this->params);
        $r->setHeader($this->headers);
        $data = $r->delete($this->params);
        $this->assertEquals(200, $data->getStatusCode());
    }

    public function testStatusCode()
    {
        echo __METHOD__ . "\n";
        $r = new CurlRequest($this->url . 'status/400');
        $r->setQueryString($this->params);
        $r->setHeader($this->headers);
        $data = $r->get();
        $this->assertEquals(400, $data->getStatusCode());
        $r = new CurlRequest($this->url . 'status/404');
        $r->setQueryString($this->params);
        $r->setHeader($this->headers);
        $data = $r->get();
        $this->assertEquals(404, $data->getStatusCode());
    }
}

src/Request.php
<?php

/**
 * User: wmateam
 * Date: 5/14/16
 * Time: 1:59 PM
 */

namespace wmateam\curling;

class Request
{
    private $channel = null;
    private $optArray = array();
    private $optHeader = array();
    private $queryString = null;

    const X_WWW_FROM_URLENCODED = 0;
    const FORM = 1;
    const RAW_DATA = 2;

    /**
     * Request constructor.
     * @param $url
     * @throws CurlingException
     */
    public function __construct($url)
    {
        if (filter_var($url, FILTER_VALIDATE_URL) === false) {
            throw new CurlingException('$url');
        }
        $this->channel = curl_init();
        $this->optArray[CURLOPT_URL] = $url;
        $this->optArray[CURLOPT_RETURNTRANSFER] = true;
        $this->optArray[CURLOPT_VERBOSE] = false;
        $this->optArray[CURLOPT_HEADER] = true;
        $this->optArray[CURLOPT_FOLLOWLOCATION] = true;
        $this->optArray[CURLOPT_ENCODING] = '';
        $this->optArray[CURLOPT_MAXREDIRS] = 10;
        $this->optArray[CURLOPT_TIMEOUT] = 30;
        $this->optArray[CURLOPT_HTTP_VERSION] = CURL_HTTP_VERSION_1_1;

    }

    /**
     * Set header params
     * @param array $header
     */
    public function setHeader($header)
    {
        array_push($this->optHeader, $header);
    }

    /**
     * @param array $data
     * @throws CurlingException
     */
    public function setQueryString($data = null)
    {
        $this->validateFields($data);
        $this->queryString = http_build_query($data);
    }

    /**
     * @return Response
     * @throws CurlingException
     */
    public function get()
    {
        $this->optArray[CURLOPT_CUSTOMREQUEST] = "GET";
        return $this->result();
    }

    /**
     * @param array $data
     * @param int $type
     * @return Response
     * @throws CurlingException
     */
    public function post($data = null, $type = self::FORM)
    {
        $this->optArray[CURLOPT_POST] = true;
        $this->optArray[CURLOPT_CUSTOMREQUEST] = "POST";
        $this->addParams($type, $data);
        return $this->result();
    }

    /**
     * @param array $data
     * @param int $type
     * @return Response
     * @throws CurlingException
     */
    public function put($data = null, $type = self::FORM)
    {
        $this->optArray[CURLOPT_CUSTOMREQUEST] = "PUT";
        $this->addParams($type, $data);
        return $this->result();
    }

    /**
     * @param array $data
     * @param int $type
     * @return Response
     * @throws CurlingException
     */
    public function patch($data = null, $type = self::FORM)
    {
        $this->optArray[CURLOPT_CUSTOMREQUEST] = "PATCH";
        $this->addParams($type, $data);
        return $this->result();
    }

    /**
     * @param array $data
     * @param int $type
     * @return Response
     * @throws CurlingException
     */
    public function delete($data = null, $type = self::FORM)
    {
        $this->optArray[CURLOPT_CUSTOMREQUEST] = "DELETE";
        $this->addParams($type, $data);
        return $this->result();
    }

    /**
     * @param $data
     * @param bool $isString
     * @throws CurlingException
     */
    private function validateFields($data, $isString = false)
    {
        if ($data != null) {
            if ($isString) {
                if (!is_string($data)) {
                    throw new CurlingException('(string)$postData', CurlingException::INVALID_TYPE);
                }
                return;
            }
            if (!is_array($data)) {
                throw new CurlingException('(array)$postData', CurlingException::INVALID_TYPE);
            }
        }
    }

    /**
     * @param $type
     * @param $data
     * @throws CurlingException
     */
    private function addParams($type, $data)
    {
        $isString = false;
        if ($type == self::RAW_DATA)
            $isString = true;


        if ($type == self::X_WWW_FROM_URLENCODED)
            array_push($this->optHeader, 'content-type: application/x-www-form-urlencoded');

        if ($type == self::RAW_DATA)
            array_push($this->optHeader, 'content-type: text/plain');
        $this->optArray[CURLOPT_HTTPHEADER] = $this->optHeader;
        $this->validateFields($data, $isString);
        $this->optArray[CURLOPT_POSTFIELDS] = $data;
    }

    /**
     * @return Response
     * @throws CurlingException
     */
    private function result()
    {
        try {
            if ($this->queryString != null)
                $this->optArray[CURLOPT_URL] = $this->optArray[CURLOPT_URL] . '?' . $this->queryString;
            curl_setopt_array($this->channel, $this->optArray);
            return new Response(curl_exec($this->channel), curl_getinfo($this->channel));
        } catch (\Exception $e) {
            throw new CurlingException($e->getMessage(), $e->getCode(), $e);
        }
    }

    /**
     * @return string
     */
    public function getErrors()
    {
        return curl_error($this->channel);
    }

    function __destruct()
    {
        curl_close($this->channel);
    }


}

src/Response.php
<?php
/**
 * User: wmateam
 * Date: 5/15/16
 * Time: 2:04 PM
 */

namespace wmateam\curling;


class Response
{
    private $result, $info;

    /**
     * Response constructor.
     * @param $result
     * @param $info
     */
    public function __construct($result, $info)
    {
        $this->result = $result;
        $this->info = $info;
    }

    public function getHeader()
    {
        $header_size = $this->getHeaderSize();
        return substr($this->result, 0, $header_size);
    }

    public function getBody()
    {
        return substr($this->result, $this->getHeaderSize());
    }

    public function getUrl()
    {
        return $this->info['url'];
    }

    public function getContentType()
    {
        return $this->info['content_type'];
    }

    public function getStatusCode()
    {
        return $this->info['http_code'];
    }

    public function getHeaderSize()
    {
        return $this->info['header_size'];
    }

    public function getRequestSize()
    {
        return $this->info['request_size'];
    }

    public function getFileTime()
    {
        return $this->info['filetime'];
    }

    public function getSslVerifyResult()
    {
        return $this->info['ssl_verify_result'];
    }

    public function getRedirectCount()
    {
        return $this->info['redirect_count'];
    }

    public function getRequestTime()
    {
        return $this->info['total_time'];
    }

    public function getSizeUpload()
    {
        return $this->info['size_upload'];
    }

    public function getSizeDownload()
    {
        return $this->info['size_download'];
    }

    public function getSpeedDownload()
    {
        return $this->info['speed_download'];
    }

    public function getSpeedUpload()
    {
        return $this->info['speed_upload'];
    }

    public function getRedirectUrl()
    {
        return $this->info['redirect_url'];
    }

}

src/HttpRequest.php
<?php

/**
 * User: wmateam
 * Date: 5/14/16
 * Time: 1:59 PM
 */

namespace wmateam\curling;

class HttpRequest
{
    private $channel = null;
    private $optArray = array();
    private $optHeader = array();
    private $queryString = null;

    const X_WWW_FROM_URLENCODED = 0;
    const FORM = 1;
    const RAW_DATA = 2;

    /**
     * Request constructor.
     * @param $url
     * @throws CurlingException
     */
    public function __construct($url)
    {
        if (!filter_var($url, FILTER_VALIDATE_URL) === false) {

        } else {
            throw new CurlingException('$url');
        }
        $this->channel = curl_init();
        $this->optArray[CURLOPT_URL] = $url;
        $this->optArray[CURLOPT_RETURNTRANSFER] = true;
        $this->optArray[CURLOPT_VERBOSE] = false;
        $this->optArray[CURLOPT_HEADER] = true;
        $this->optArray[CURLOPT_FOLLOWLOCATION] = true;
        $this->optArray[CURLOPT_ENCODING] = '';
        $this->optArray[CURLOPT_MAXREDIRS] = 10;
        $this->optArray[CURLOPT_TIMEOUT] = 30;
        $this->optArray[CURLOPT_HTTP_VERSION] = CURL_HTTP_VERSION_1_1;
        $this->optArray[CURLOPT_USERAGENT] = 'wmateam-curling/0.5.0';
    }

    /**
     * @return HttpResponse
     * @throws CurlingException
     */
    public function get()
    {
        $this->optArray[CURLOPT_CUSTOMREQUEST] = "GET";
        return $this->result();
    }

    /**
     * @param array $data
     * @param int $type
     * @param bool $isJSON
     * @return HttpResponse
     * @throws CurlingException
     */
    public function post($data = null, $type = self::FORM, $isJSON = false)
    {
        $this->optArray[CURLOPT_POST] = true;
        $this->optArray[CURLOPT_CUSTOMREQUEST] = "POST";
        $this->addParams($type, $data, $isJSON);
        return $this->result();
    }

    /**
     * @param array $data
     * @param int $type
     * @param bool $isJSON
     * @return HttpResponse
     * @throws CurlingException
     */
    public function put($data = null, $type = self::FORM, $isJSON = false)
    {
        $this->optArray[CURLOPT_CUSTOMREQUEST] = "PUT";
        $this->addParams($type, $data, $isJSON);
        return $this->result();
    }

    /**
     * @param array $data
     * @param int $type
     * @param bool $isJSON
     * @return HttpResponse
     * @throws CurlingException
     */
    public function patch($data = null, $type = self::FORM, $isJSON = false)
    {
        $this->optArray[CURLOPT_CUSTOMREQUEST] = "PATCH";
        $this->addParams($type, $data, $isJSON);
        return $this->result();
    }

    /**
     * @param array $data
     * @param int $type
     * @param bool $isJSON
     * @return HttpResponse
     * @throws CurlingException
     */
    public function delete($data = null, $type = self::FORM, $isJSON = false)
    {
        $this->optArray[CURLOPT_CUSTOMREQUEST] = "DELETE";
        $this->addParams($type, $data, $isJSON);
        return $this->result();
    }

    public function setAuthentication($username, $password)
    {
        $this->optArray[CURLOPT_USERPWD] = $username . ':' . $password;
        $this->optArray[CURLOPT_UNRESTRICTED_AUTH] = true;
    }


    public function setProxy($proxy, $username = null, $password = null)
    {
        if ($proxy === '') {
            throw new CurlingException('$proxy', CurlingException::INVALID_TYPE);
        }

        $this->optArray[CURLOPT_PROXY] = $proxy;
        $this->optArray[CURLOPT_PROXYUSERPWD] = $username . ':' . $password;
    }

    /**
     * Set header params
     * @param string $header
     */
    public function setHeader($header)
    {
        array_push($this->optHeader, $header);
    }

    /**
     * @param array $data
     * @throws CurlingException
     */
    public function setQueryString($data = null)
    {
        $this->validateFields($data);
        $this->queryString = http_build_query($data);
    }

    public function setUserAgent($agent)
    {
        if (!is_string($agent))
            throw new CurlingException('(string)$agent', CurlingException::INVALID_TYPE);
        $this->userAgent($agent);
    }

    private function userAgent($agent)
    {
        $this->optArray[CURLOPT_USERAGENT] = $agent;
    }

    /**
     * @param $data
     * @param bool $isString
     * @throws CurlingException
     */
    private function validateFields($data, $isString = false)
    {
        if ($data != null) {
            if ($isString) {
                if (!is_string($data)) {
                    throw new CurlingException('(string)$postData', CurlingException::INVALID_TYPE);
                }
                return;
            }
            if (!is_array($data)) {
                throw new CurlingException('(array)$postData', CurlingException::INVALID_TYPE);
            }
        }
    }

    /**
     * @param $type
     * @param $data
     * @param bool $isJSON
     * @throws CurlingException
     */
    private function addParams($type, $data, $isJSON = false)
    {
        $isString = false;


        if ($type == self::X_WWW_FROM_URLENCODED)
            $this->setHeader('content-type: application/x-www-form-urlencoded');

        if ($type == self::RAW_DATA) {
            $isString = true;
            if ($isJSON)
                $this->setHeader('content-type: application/json');
            else
                $this->setHeader('content-type: text/plain');
        }

        $this->validateFields($data, $isString);
        $this->optArray[CURLOPT_POSTFIELDS] = $data;
    }

    /**
     * @return HttpResponse
     * @throws CurlingException
     */
    private function result()
    {
        try {
            if ($this->queryString != null)
                $this->optArray[CURLOPT_URL] = $this->optArray[CURLOPT_URL] . '?' . $this->queryString;
            $this->optArray[CURLOPT_HTTPHEADER] = $this->optHeader;
            curl_setopt_array($this->channel, $this->optArray);
            return new HttpResponse(curl_exec($this->channel), curl_getinfo($this->channel));
        } catch (\Exception $e) {
            throw new CurlingException($e->getMessage(), $e->getCode(), $e);
        }
    }

    /**
     * @return string
     */
    public function getErrors()
    {
        return curl_error($this->channel);
    }

    function __destruct()
    {
        curl_close($this->channel);
    }


}

src/CurlResponse.php
<?php
/**
 * User: wmateam
 * Date: 5/15/16
 * Time: 2:04 PM
 */

namespace wmateam\curling;


class CurlResponse
{
    private $result, $info, $isJson;

    /**
     * Response constructor.
     *
     * @param $result
     * @param $info
     */
    public function __construct($result, $info)
    {
        $this->result = $result;
        $this->info = $info;
        $this->isJson = preg_match('/(application\/json)/i', $this->getContentType());
    }

    /**
     * return response header
     *
     * @return string
     */
    public function getHeader()
    {
        $header_size = $this->getHeaderSize();
        return substr($this->result, 0, $header_size);
    }

    /**
     * retrieve response body
     *
     * @return string
     */
    public function getBody()
    {
        return substr($this->result, $this->getHeaderSize());
    }

    /**
     * return url
     *
     * @return string
     */
    public function getUrl()
    {
        return $this->info['url'];
    }

    /**
     * return response content type
     *
     * @return mixed
     */
    public function getContentType()
    {
        return $this->info['content_type'];
    }

    /**
     * return status code
     *
     * @return integer
     */
    public function getStatusCode()
    {
        return $this->info['http_code'];
    }

    /**
     * return header size
     *
     * @return integer
     */
    public function getHeaderSize()
    {
        return $this->info['header_size'];
    }

    /**
     * return request size
     *
     * @return integer
     */
    public function getRequestSize()
    {
        return $this->info['request_size'];
    }

    /**
     * return file time
     *
     * @return integer
     */
    public function getFileTime()
    {
        return $this->info['filetime'];
    }

    /**
     * return ssl verification result
     *
     * @return integer
     */
    public function getSslVerifyResult()
    {
        return $this->info['ssl_verify_result'];
    }

    /**
     * return redirection count
     *
     * @return integer
     */
    public function getRedirectCount()
    {
        return $this->info['redirect_count'];
    }

    /**
     * return request time
     * @return double
     */
    public function getRequestTime()
    {
        return $this->info['total_time'];
    }

    /**
     * return upload size
     * @return double
     */
    public function getSizeUpload()
    {
        return $this->info['size_upload'];
    }

    /**
     * return download size
     * @return double
     */
    public function getSizeDownload()
    {
        return $this->info['size_download'];
    }

    /**
     * return download speed
     * @return double
     */
    public function getSpeedDownload()
    {
        return $this->info['speed_download'];
    }

    /**
     * return upload speed
     * @return double
     */
    public function getSpeedUpload()
    {
        return $this->info['speed_upload'];
    }

    /**
     * return redirect url
     * @return string
     */
    public function getRedirectUrl()
    {
        return $this->info['redirect_url'];
    }

    /**
     * return response isJson ?
     * @return bool
     */
    public function isJson()
    {
        return $this->isJson > 0;
    }

    /**
     * return json Response
     * @param bool $asArray return as Array?
     * @return \stdClass|array
     */
    public function getJson($asArray = false)
    {
        if ($this->isJson())
            return json_decode($this->getBody(), $asArray);
        return null;
    }
}

src/HttpResponse.php
<?php
/**
 * User: wmateam
 * Date: 5/15/16
 * Time: 2:04 PM
 */

namespace wmateam\curling;


class HttpResponse
{
    private $result, $info;

    /**
     * Response constructor.
     * @param $result
     * @param $info
     */
    public function __construct($result, $info)
    {
        $this->result = $result;
        $this->info = $info;
    }

    public function getHeader()
    {
        $header_size = $this->getHeaderSize();
        return substr($this->result, 0, $header_size);
    }

    public function getBody()
    {
        return substr($this->result, $this->getHeaderSize());
    }

    public function getUrl()
    {
        return $this->info['url'];
    }

    public function getContentType()
    {
        return $this->info['content_type'];
    }

    public function getStatusCode()
    {
        return $this->info['http_code'];
    }

    public function getHeaderSize()
    {
        return $this->info['header_size'];
    }

    public function getRequestSize()
    {
        return $this->info['request_size'];
    }

    public function getFileTime()
    {
        return $this->info['filetime'];
    }

    public function getSslVerifyResult()
    {
        return $this->info['ssl_verify_result'];
    }

    public function getRedirectCount()
    {
        return $this->info['redirect_count'];
    }

    public function getRequestTime()
    {
        return $this->info['total_time'];
    }

    public function getSizeUpload()
    {
        return $this->info['size_upload'];
    }

    public function getSizeDownload()
    {
        return $this->info['size_download'];
    }

    public function getSpeedDownload()
    {
        return $this->info['speed_download'];
    }

    public function getSpeedUpload()
    {
        return $this->info['speed_upload'];
    }

    public function getRedirectUrl()
    {
        return $this->info['redirect_url'];
    }

}

src/CurlingException.php
<?php
/**
 * Created by PhpStorm.
 * User: mohammad
 * Date: 5/14/16
 * Time: 2:47 PM
 */

namespace wmateam\curling;


use Exception;

class CurlingException extends Exception
{
    const INVALID_TYPE = 0;
    const INVALID_URL = 0;

    public function __construct($objectName, $code = self::INVALID_TYPE, Exception $previous = null)
    {
        switch ($code) {
            case self::INVALID_TYPE:
                parent::__construct('Invalid type for ' . $objectName, $code);
                break;
            case self::INVALID_URL:
                parent::__construct('Invalid url for ' . $objectName, $code);
                break;
            default:
                parent::__construct($objectName, $code, $previous);
                break;
        }
    }

    public function toString()
    {
        return __CLASS__ . ": [{$this->code}]: {$this->message}\n";
    }

    public function toArray()
    {
        return array(
            'Message' => $this->getMessage(),
            'File' => $this->getFile(),
            'Line' => $this->getLine(),
            'Trace' => $this->getTraceAsString()
        );
    }


    function __destruct()
    {
        // TODO: Implement __destruct() method.
    }


}

src/CurlRequest.php
<?php

/**
 * User: wmateam
 * Date: 5/14/16
 * Time: 1:59 PM
 */

namespace wmateam\curling;

class CurlRequest
{
    private $channel = null;
    private $optArray = array();
    private $optHeader = array();
    private $queryString = null;

    const X_WWW_FROM_URLENCODED = 0;
    const FORM = 1;
    const RAW_DATA = 2;

    /**
     * Request constructor.
     * @param $url
     * @throws CurlingException
     */
    public function __construct($url)
    {
        if (!filter_var($url, FILTER_VALIDATE_URL) === false) {

        } else {
            throw new CurlingException('$url');
        }
        $this->channel = curl_init();
        $this->optArray[CURLOPT_URL] = $url;
        $this->optArray[CURLOPT_RETURNTRANSFER] = true;
        $this->optArray[CURLOPT_VERBOSE] = false;
        $this->optArray[CURLOPT_HEADER] = true;
        $this->optArray[CURLOPT_FOLLOWLOCATION] = true;
        $this->optArray[CURLOPT_ENCODING] = '';
        $this->optArray[CURLOPT_MAXREDIRS] = 10;
        $this->optArray[CURLOPT_TIMEOUT] = 30;
        $this->optArray[CURLOPT_HTTP_VERSION] = CURL_HTTP_VERSION_1_1;
        $this->optArray[CURLOPT_USERAGENT] = 'wmateam-curling/0.7.5';
    }

    /**
     * @return CurlResponse
     * @throws CurlingException
     */
    public function get()
    {
        $this->optArray[CURLOPT_CUSTOMREQUEST] = "GET";
        return $this->result();
    }

    /**
     * @param array $data
     * @param int $type
     * @param bool $isJSON
     * @return CurlResponse
     * @throws CurlingException
     */
    public function post($data = null, $type = self::FORM, $isJSON = false)
    {
        $this->optArray[CURLOPT_POST] = true;
        $this->optArray[CURLOPT_CUSTOMREQUEST] = "POST";
        $this->addParams($type, $data, $isJSON);
        return $this->result();
    }

    /**
     * @param array $data
     * @param int $type
     * @param bool $isJSON
     * @return CurlResponse
     * @throws CurlingException
     */
    public function put($data = null, $type = self::FORM, $isJSON = false)
    {
        $this->optArray[CURLOPT_CUSTOMREQUEST] = "PUT";
        $this->addParams($type, $data, $isJSON);
        return $this->result();
    }

    /**
     * @param array $data
     * @param int $type
     * @param bool $isJSON
     * @return CurlResponse
     * @throws CurlingException
     */
    public function patch($data = null, $type = self::FORM, $isJSON = false)
    {
        $this->optArray[CURLOPT_CUSTOMREQUEST] = "PATCH";
        $this->addParams($type, $data, $isJSON);
        return $this->result();
    }

    /**
     * @param array $data
     * @param int $type
     * @param bool $isJSON
     * @return CurlResponse
     * @throws CurlingException
     */
    public function delete($data = null, $type = self::FORM, $isJSON = false)
    {
        $this->optArray[CURLOPT_CUSTOMREQUEST] = "DELETE";
        $this->addParams($type, $data, $isJSON);
        return $this->result();
    }

    public function setAuthentication($username, $password)
    {
        $this->optArray[CURLOPT_USERPWD] = $username . ':' . $password;
        $this->optArray[CURLOPT_UNRESTRICTED_AUTH] = true;
    }


    public function setProxy($proxy, $username = null, $password = null)
    {
        if ($proxy === '') {
            throw new CurlingException('$proxy', CurlingException::INVALID_TYPE);
        }

        $this->optArray[CURLOPT_PROXY] = $proxy;
        $this->optArray[CURLOPT_PROXYUSERPWD] = $username . ':' . $password;
    }

    /**
     * Set header params
     * @param string $header
     */
    public function setHeader($header)
    {
        array_push($this->optHeader, $header);
    }

    /**
     * @param array $data
     * @throws CurlingException
     */
    public function setQueryString($data = null)
    {
        $this->validateFields($data);
        $this->queryString = http_build_query($data);
    }

    public function setUserAgent($agent)
    {
        if (!is_string($agent))
            throw new CurlingException('(string)$agent', CurlingException::INVALID_TYPE);
        $this->userAgent($agent);
    }

    private function userAgent($agent)
    {
        $this->optArray[CURLOPT_USERAGENT] = $agent;
    }

    /**
     * @param $data
     * @param bool $isString
     * @throws CurlingException
     */
    private function validateFields($data, $isString = false)
    {
        if ($data != null) {
            if ($isString) {
                if (!is_string($data)) {
                    throw new CurlingException('(string)$postData', CurlingException::INVALID_TYPE);
                }
                return;
            }
            if (!is_array($data)) {
                throw new CurlingException('(array)$postData', CurlingException::INVALID_TYPE);
            }
        }
    }

    /**
     * @param $type
     * @param $data
     * @param bool $isJSON
     * @throws CurlingException
     */
    private function addParams($type, $data, $isJSON = false)
    {
        $isString = false;


        if ($type == self::X_WWW_FROM_URLENCODED) {
            $this->setHeader('content-type: application/x-www-form-urlencoded');
            $data = http_build_query($data);
        }

        if ($type == self::RAW_DATA) {
            $isString = true;
            if ($isJSON)
                $this->setHeader('content-type: application/json');
            else
                $this->setHeader('content-type: text/plain');
        }

        $this->validateFields($data, $isString);
        $this->optArray[CURLOPT_POSTFIELDS] = $data;
    }

    /**
     * @return CurlResponse
     * @throws CurlingException
     */
    private function result()
    {
        try {
            if ($this->queryString != null)
                $this->optArray[CURLOPT_URL] = $this->optArray[CURLOPT_URL] . '?' . $this->queryString;
            $this->optArray[CURLOPT_HTTPHEADER] = $this->optHeader;
            curl_setopt_array($this->channel, $this->optArray);
            return new CurlResponse(curl_exec($this->channel), curl_getinfo($this->channel));
        } catch (\Exception $e) {
            throw new CurlingException($e->getMessage(), $e->getCode(), $e);
        }
    }

    /**
     * @return string
     */
    public function getErrors()
    {
        return curl_error($this->channel);
    }

    function __destruct()
    {
        curl_close($this->channel);
    }


}